db: Added schema and data source for table `system.clients'

jul-stas opened the pull request:
This change introduces `system.clients` table, which provides information about CQL clients connected.

PK is the client's IP address, CK consists of outgoing `port` number and `client_type` (which will be extended in future to thrift/alternator/redis). Table supplies also `shard_id` and `username`. Other columns, like `connection_stage`, `driver_name`, `driver_version`..., are currently empty but exist for C* compatibility and future use.

This is an ordinary table (i.e. non-virtual) and it's updated upon accepting connections. This is also why C*'s column `request_count` was not introduced. In case of abrupt DB stop, the table should not persist, so it's being truncated on startup.
Resolves #4820

avikivity commented on db/system_keyspace_clients.cc:
```diff
@@ -0,0 +1,106 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
```
Since this is new code, it shouldn't be licensed to Apache.

avikivity commented on db/system_keyspace_clients.cc:
```diff
@@ -0,0 +1,106 @@
+ */
+
+/*
+ * Modified by ScyllaDB
```
This is only needed on modified Apache code, not new code.

avikivity commented on db/system_keyspace_clients.cc:
```diff
@@ -0,0 +1,106 @@
+                {"driver_version", utf8_type},
+                {"hostname", utf8_type},
+                {"protocol_version", int32_type},
+                {"request_count", long_type},
```
How can we implement request count with a table? This is updated continuously.

jul-stas replied:
As discussed below, I removed this column

avikivity commented on transport/server.cc:
```diff
@@ -236,23 +239,44 @@ cql_server::do_accepts(int which, bool keepalive, socket_address server_addr) {
 
+future<>
+cql_server::advertise_new_connection(shared_ptr<connection> conn) {
+    db::system_keyspace::client_data cd;
```
This is introducing a dependency between system_keyspace and transport. The way to break the dependency is to create an interface/implementation pair and have transport and system_keyspace only talk to the interface, not each other. main.cc can mediate between them.

jul-stas replied:
> The problems you listed don't worry me
Good to hear that, but please note that now `Scylla version ... initialization completed.` can no longer be the last line in startup logs (because of compaction logs covering it). It doesn't hurt, but also doesn't look as "inviting". Don't you think it should be dealt with?

`[shard 0] init - Scylla version 666.development-0.20191029.767f636eb initialization completed.
[shard 0] compaction - Compacted 2 sstables to [tmp/system/truncated-38c19fd0fb863310a4b70d0cc66628aa/mc-102-big-Data.db:level=0, ]. 82010 bytes to 76766 (~93% of original) in 294ms = 0.25MB/s. ~256 total partitions merged to 2.
[shard 0] compaction - Compacting [tmp/system/truncated-38c19fd0fb863310a4b70d0cc66628aa/mc-99-big-Data.db:level=0, tmp/system/truncated-38c19fd0fb863310a4b70d0cc66628aa/mc-102-big-Data.db:level=0, ]
[shard 0] compaction - Compacted 2 sstables to [tmp/system/truncated-38c19fd0fb863310a4b70d0cc66628aa/mc-105-big-Data.db:level=0, ]. 82002 bytes to 76768 (~93% of original) in 290ms = 0.25MB/s. ~256 total partitions merged to 2.
`

jul-stas replied:
I decoupled them, but in a simpler way. The code that announces connection/disconnection was stateless, so it was enough just to shuffle things around.

avikivity commented:
The problems you listed don't worry me, but request_count does. It's an important metric too, since it allows the administrator to detect idle/dead connections and wild connections.

We might decide not to implement request_count for now (and drop it from the table) and add it later when we have better virtual table infrastructure. @glommer will you be able to survive without request_count?

glommer commented:
> The problems you listed don't worry me, but request_count does. It's an important metric too, since it allows the administrator to detect idle/dead connections and wild connections.
> 
> We might decide not to implement request_count for now (and drop it from the table) and add it later when we have better virtual table infrastructure. @glommer will you be able to survive without request_count?

I believe so.

This is the kind of thing that is better suited for the metric, not table domain anyway

avikivity commented:
> This is the kind of thing that is better suited for the metric, not table domain anyway

What do you mean? 

If you mean this should be reported by the metrics API, we're not going to generate a separate metric for each connection.

avikivity commented on db/system_keyspace_clients.cc:
```diff
@@ -0,0 +1,106 @@
+                {"ssl_cipher_suite", utf8_type},
+                {"ssl_enabled", boolean_type},
+                {"ssl_protocol", utf8_type},
+                {"username", utf8_type}
```
Need to add shard here so the user knows which shard terminates the connection.

jul-stas replied:
done

avikivity commented on db/system_keyspace_clients.cc:
```diff
@@ -0,0 +1,106 @@
+            // partition key
+            {{"address", inet_addr_type},},
+            // clustering key
+            {{"port", int32_type}},
```
Need to add protocol to the clustering key (as the first component) so that we can list alternator and thrift clients too (no need to add them now)

jul-stas replied:
done

glommer commented:
> > This is the kind of thing that is better suited for the metric, not table domain anyway
> 
> What do you mean?
> 
> If you mean this should be reported by the metrics API, we're not going to generate a separate metric for each connection.

Things that are fast moving like this and increasing over time are more useful if graphed, so yes, that's what I mean.

Now, I don't think we strictly need this at all, so don't understand my statement as "we must do this in the metric layer". Rather : if we were to do this, it would be more useful in the metrics layer anyway.

If I understood the PR correctly, this field adds problems, since it is a fast moving field in a real table, *and* we would only be able to realize its full usefulness in the metric layer anyway. Since we don't want that either: drop it.

jul-stas commented:
> The problems you listed don't worry me

Good to hear that, but please note that now `Scylla version ... initialization completed.` can no longer be the last line in startup logs (because of compaction logs covering it). It doesn't hurt, but also doesn't look as "inviting". Don't you think it should be dealt with?

`[shard 0] init - Scylla version 666.development-0.20191029.767f636eb initialization completed.
[shard 0] compaction - Compacted 2 sstables to [tmp/system/truncated-38c19fd0fb863310a4b70d0cc66628aa/mc-102-big-Data.db:level=0, ]. 82010 bytes to 76766 (~93% of original) in 294ms = 0.25MB/s. ~256 total partitions merged to 2.
[shard 0] compaction - Compacting [tmp/system/truncated-38c19fd0fb863310a4b70d0cc66628aa/mc-99-big-Data.db:level=0, tmp/system/truncated-38c19fd0fb863310a4b70d0cc66628aa/mc-102-big-Data.db:level=0, ]
[shard 0] compaction - Compacted 2 sstables to [tmp/system/truncated-38c19fd0fb863310a4b70d0cc66628aa/mc-105-big-Data.db:level=0, ]. 82002 bytes to 76768 (~93% of original) in 290ms = 0.25MB/s. ~256 total partitions merged to 2.
`

jul-stas commented on connection_notifier.cc:
```diff
@@ -0,0 +1,59 @@
+    
+    return db::execute_cql(req,
+            std::move(cd.ip), to_string(cd.clit), cd.port, cd.shard_id,
+            cd.protocol_version.has_value() ? data_value(*cd.protocol_version) : data_value::make_null(int32_type),
```
I'm unhappy about introducing tombstone here (that's what will happen, right?). Any ideas are welcome.

avikivity replied:
There is no way around it. However, you can set gc_grace_second = 0 in the schema to reduce the time the tombstone hangs around (this is safe since this is a local table and will never be repaired)

jul-stas commented:
@avikivity Any more remarks or can we merge this?

avikivity commented on db/system_keyspace_clients.cc:
```diff
@@ -0,0 +1,98 @@
+            // partition key
+            {{"address", inet_addr_type}},
+            // clustering key
+            {{"client_type", utf8_type}, {"port", int32_type}},
```
I think we should swap these two, so a `SELECT * FROM system.clients WHERE address = '1.2.3.4' AND port = 1234` works without ALLOW FILTERING. Usually there will be only one client type, so selecting by client type is less interesting.

avikivity commented on db/system_keyspace_clients.cc:
```diff
@@ -0,0 +1,98 @@
+    thread_local auto clients = [] {
+        schema_builder builder(make_lw_shared(schema(generate_legacy_id(NAME, CLIENTS), NAME, CLIENTS,
+            // partition key
+            {{"address", inet_addr_type}},
```
Let's rename to client_address and client_port to avoid ambiguity

avikivity commented on db/system_keyspace_clients.cc:
```diff
@@ -0,0 +1,98 @@
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
```
Missing copyright statement.  A license isn't enough.

avikivity commented on db/system_keyspace_clients.cc:
```diff
@@ -0,0 +1,98 @@
+            // comment
+            "list of connected CQL clients"
+            )));
+        builder.set_gc_grace_seconds(0);
```
Oh, you already did that, so don't worry about tombstones.

avikivity commented on db/system_keyspace_clients.cc:
```diff
@@ -0,0 +1,98 @@
+}
+
+future<> insert_client(client_data cd) {
+    const thread_local sstring req
```
static is enough, no need for thread_local for const objects.

avikivity commented on db/system_keyspace_clients.cc:
```diff
@@ -0,0 +1,98 @@
+            cd.username.value_or("anonymous")).discard_result();
+}
+
+future<> remove_client(gms::inet_address addr, client_type clit, int port) {
```
Maybe rename `clit` so we don't get marked as spam in the email lists.

avikivity commented on db/system_keyspace_clients.hh:
```diff
@@ -0,0 +1,60 @@
+enum class client_type {
+    cql = 0,
+    thrift,
+    alternator
```
A trailing comma is customary as a curtesy to the next one to add an element

avikivity commented on db/system_keyspace_clients.cc:
```diff
@@ -0,0 +1,98 @@
+future<> clear_clientlist() {
+    const thread_local sstring req
+            = format("TRUNCATE system.{};", CLIENTS);
+    slogger.info("Clearing non-persistent CF: `{}'", req);
```
This is too much detail for info level.

avikivity commented on transport/server.cc:
```diff
@@ -236,23 +238,49 @@ cql_server::do_accepts(int which, bool keepalive, socket_address server_addr) {
+                        clogger.debug("connection error: {}", std::current_exception());
+                    }
+                }).finally([this, conn] {
+                    --_connections;
```
If advertise_new_connection() fails, then we won't decrement _connections, even though we incremented it previously.

We also lose the finally() section of connection::close().

avikivity commented on transport/server.cc:
```diff
@@ -236,23 +238,49 @@ cql_server::do_accepts(int which, bool keepalive, socket_address server_addr) {
+            return advertise_new_connection(conn).then([this, conn] {
+                // Move connection to the background, monitor for lifetime and errors.
+                (void)conn->process().then_wrapped([this, conn] (future<> f) {
+                    try {
```
I think the (void) should be moved to the caller. The accept loop shouldn't wait for advertise_new_connection(), and shouldn't fail if a_n_c() fails.

avikivity commented on transport/server.cc:
```diff
@@ -236,23 +238,49 @@ cql_server::do_accepts(int which, bool keepalive, socket_address server_addr) {
         }).handle_exception([] (auto ep) {
             clogger.debug("accept failed: {}", ep);
-            return stop_iteration::no;
+            return make_ready_future<stop_iteration>(stop_iteration::no);
```
Probably unnecessary.

avikivity commented on connection_notifier.cc:
```diff
@@ -0,0 +1,59 @@
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
```
Missing copyright.

avikivity commented on db/system_keyspace.cc:
```diff
@@ -543,6 +542,46 @@ static schema_ptr large_cells() {
+            {{"address", inet_addr_type}},
+            // clustering key
+            {{"client_type", utf8_type}, {"port", int32_type}},
+            // regular columns
```
Why am I seeing this twice?

avikivity replied:
If `address` and `port` come from cassandra, we should keep the names.

jul-stas replied:
Yes, column names come from C*, so I left them as is.

avikivity commented on db/system_keyspace_clients.cc:
```diff
@@ -1,98 +0,0 @@
-/*
- * This file is part of Scylla.
- *
```
I don't understand why I'm reviewing a file being added and deleted. 

Patchsets should be constructed in a way that is friendly to reviewers, not make them do extra work be reviewing everything several times and trying to reverse engineer what happened.

jul-stas replied:
My apologies - I thought that reviewers inspect the "Changed files" tab, while individual commits represent the "historical thought process" behind PR. Spurious changesets are now squashed away.

jul-stas commented:
@avikivity I fixed the code according to your comment, could you please have another look?

avikivity commented:
What are the changes relative to the previous version?

jul-stas commented:
Changes in 5a34ce6:
- rearranged call to `advertise_new_connection` in `transport/server.cc` to this: https://github.com/scylladb/scylla/blob/5a34ce6e659953c3111f43439fd90e76450809ca/transport/server.cc#L242
Future that is tossed into background now contains `{advertise_new,unadvertise}_connection` and counter increment/decrement. The middlemost lambda doesn't throw, so I decided to put decrement and unadvertise into top-level `.then()` continuation.
- swapped the order of columns `client_type` and `port` in `system_keyspace.cc:564` (also in query strings in `connection_notifier.cc:{43,54}`)
- added copyright notice to `connection_notifier.*:1`
- made query strings `static` instead of `thread_local` in `connection_notifier.cc:{42,53,60}`
- renamed `client_type clit` to `ct` in all occurences (oops, sorry)
- lowered log level to trace in `transport/server.cc:{268,276}`
- removed `make_ready_future` from `transport/server.cc:257`
- and most importantly: added trailing comma to `enum class client_type` in `connection_notifier.hh:27`

avikivity commented on main.cc:
```diff
@@ -67,6 +67,8 @@
 #include "gms/feature_service.hh"
 #include "distributed_loader.hh"
 #include "cql3/cql_config.hh"
+#include "serializer.hh"
```
Is this needed?

jul-stas replied:
My bad. Removed.

avikivity commented on main.cc:
```diff
@@ -1117,6 +1122,7 @@ int main(int ac, char** av) {
                     return logalloc::shard_tracker().enable_abort_on_bad_alloc();
                 }).get();
             }
+
```
And this?

jul-stas replied:
My bad. Removed.

avikivity commented on service/client_state.hh:
```diff
@@ -68,6 +68,9 @@ public:
     };
 
+    /// Conversion to string according to C* convention in `system.clients`.
+    sstring to_cassandra_string(auth_state as);
```
This should either be static, or use client_state::_auth_state

jul-stas replied:
Or, even better, removed. (anyway `auth_state` cannot be handled properly at this point, because it changes after connection)

avikivity commented on transport/server.cc:
```diff
@@ -237,24 +239,44 @@ cql_server::do_accepts(int which, bool keepalive, socket_address server_addr) {
-            (void)conn->process().then_wrapped([this, conn] (future<> f) {
+            (void)advertise_new_connection(conn).then([this, conn] {
+                ++_connections;
+                // Monitor connection's lifetime and errors in the background.
```
The comment should not be moved, since it (tries to) explain the cast to void. In general casting a future to void is a no-no so we require an explanation about each call site.

I think the ++_connections should remain in its place. Consider a very slow advertise_new_connection(), we could have lots of new connections stuck there, with the connections metric not showing any change, because it is only updated after a_n_c().

jul-stas replied:
Here's my new take on this: https://github.com/jul-stas/scylla/blob/99d8441fd408c306fa09c20655487d5a5370eb8c/transport/server.cc#L240
I tried hybrid approach, with following assumptions:
- if `a_n_c()` fails exception is logged and connection proceeds normally
- if `conn->process()` fails, exception is logged and `unadvertise...()` is called anyway
- if both `conn->process()` and `unadvertise...()` fail then `nested_exception` is logged and accept() loops normally
It has its flaws (eg. unadvertise is called in the same fashion regardless of scenario) but it's the best I could think of ATM.

avikivity commented on transport/server.cc:
```diff
@@ -237,24 +239,44 @@ cql_server::do_accepts(int which, bool keepalive, socket_address server_addr) {
-                } catch (...) {
-                    clogger.debug("connection error: {}", std::current_exception());
-                }
+                return unadvertise_connection(conn);
```
I think an exception here won't be logged, since the future is dropped (cast to void). So need to log it explictly here.

(there is a last ditch logger for discarded exceptional futures, but we try not to rely on it, since it doesn't provide any context to the user).

We may want to log and ignore errors when adding connections, too, since a problem in that table can kill the node by refusing any new connections, including those that attempt to fix the table.

jul-stas replied:
please see comment above (or, even better, the code)

avikivity commented on transport/server.hh:
```diff
@@ -177,6 +178,8 @@ private:
         future<> process();
         future<> process_request();
         future<> shutdown();
+        void fill_client_data(client_data& cd) const;  // Arg here must go by ref (because fwd decl)
```
It's nicer to return a value instead of using an out-parameter.

jul-stas replied:
Done. (although that introduces one more `#include` which I tried to avoid)

avikivity replied:
Why do you need a #include? returning a forward declared class should work, no?

avikivity commented on configure.py:
```diff
@@ -470,6 +470,7 @@ def find_headers(repodir, excluded_dirs):
                 'table.cc',
                 'atomic_cell.cc',
```
Please split into two patches: infrastructure up to the new API in connection_notifier.hh, and a second patch that uses the API in transport/server.cc. This eases review by allowing the reviewer to grok the infrastructure, and then consider how it is used.

avikivity replied:
(and if we implement thrift and alternator support, these would come as further separate patches that can be reviewed by those protocols' maintainers)

jul-stas replied:
Done

jul-stas commented:
@avikivity I think it's ready for another round of review. I replied under comments.

avikivity commented on transport/server.hh:
```diff
@@ -38,6 +38,7 @@
 #include "utils/fragmented_temporary_buffer.hh"
 #include "service_permit.hh"
+#include "connection_notifier.hh"
 
```
See https://godbolt.org/z/vqA968 for an example where forward declaration is sufficient.

jul-stas replied:
I'm retarded.

avikivity commented on main.cc:
```diff
@@ -68,6 +68,7 @@
 #include "gms/feature_service.hh"
```
The prefix for the commit subject should be "transport:" rather than "db:" since it applies to the transport layer.

jul-stas replied:
done

avikivity commented on transport/server.cc:
```diff
@@ -238,14 +238,21 @@ cql_server::do_accepts(int which, bool keepalive, socket_address server_addr) {
-                    clogger.debug("connection error: {}", std::current_exception());
+                } catch(...) {
+                    clogger.debug("exception while advertising new connection: {}", std::current_exception());
                 }
```
Should be at INFO level, or we'll never see it in regular deployments. Since we don't expect an exception under normal circumstances, it won't pollute the log too much.

jul-stas replied:
Done.

avikivity commented on transport/server.cc:
```diff
@@ -238,14 +238,21 @@ cql_server::do_accepts(int which, bool keepalive, socket_address server_addr) {
+                    --_connections;
+                    return unadvertise_connection(conn);
+                }).handle_exception([] (auto ep) {
+                    clogger.debug("exception found after connection close: {}", ep);
```
Again, info level or we'll never see it.

Please avoid gratuitus generic lambdas, they're hard for IDEs to understand.

jul-stas replied:
Both done.

avikivity commented:
Looks good, I had only very minor comments.

avikivity commented on configure.py:
```diff
@@ -470,6 +470,7 @@ def find_headers(repodir, excluded_dirs):
                 'table.cc',
```
Why did you change this commit's subsystem tag? It definitely is not in the transport layer.

jul-stas replied:
My excessive need for unification. Now it's `system_keyspace:`

avikivity replied:
Please also adjust the cover letter (that goes into the merge commit). I often tweet interesting commits, so it should be updated and allow readers to understand what is going on.

avikivity replied:
To be clear, the cover letter should be stand alone and not written as a continuation from the previous attempt.

jul-stas replied:
Done!
